admin/load.php

```diff
@@ -609,18 +641,97 @@ function perflab_deactivate_plugin() {
  *
  * @since n.e.x.t
- *
- * @return void
  */
-function perflab_plugin_admin_notices() {
-	if ( 'settings_page_perflab-modules' !== get_current_screen()->id ) {
-		return;
+function perflab_install_activate_standalone_plugins_callback() {
+	if ( ! wp_verify_nonce( $_REQUEST['nonce'], 'perflab-install-activate-plugins' ) ) {
+		$status['errorMessage'] = esc_html__( 'Invalid nonce: Please refresh and try again.', 'performance-lab' );
+		wp_send_json_error( $status );
+	}
+
+	if ( ! current_user_can( 'install_plugins' ) || ! current_user_can( 'activate_plugins' ) ) {
+		$status['errorMessage'] = esc_html__( 'Sorry, you are not allowed to manage plugins for this site. Please contact the administrator.', 'performance-lab' );
+		wp_send_json_error( $status );
+	}
+
+	if ( ! function_exists( 'plugins_api' ) ) {
+		require_once ABSPATH . 'wp-admin/includes/plugin-install.php';
+	}
+	require_once ABSPATH . 'wp-admin/includes/class-wp-upgrader.php';
+	require_once ABSPATH . 'wp-admin/includes/class-wp-ajax-upgrader-skin.php';
+
+	$plugins_to_activate = perflab_get_active_modules_with_standalone_plugins();
+	$modules             = perflab_get_module_settings();
+	$plugins             = get_plugins();
+	$status              = array();
+
+	foreach ( $plugins_to_activate as $module_slug ) {
+
+		// Skip checking for already activated plugin.
+		if ( perflab_is_standalone_plugin_loaded( $module_slug ) ) {
+			continue;
+		}
+
+		$plugin_slug     = basename( $module_slug );
+		$plugin_basename = $plugin_slug . '/load.php';
+		$api             = perflab_query_plugin_info( $plugin_slug );
+
+		// Return early if plugin API return an error.
+		if ( ! $api ) {
+			$status['errorMessage'] = __( 'An unexpected error occurred. Something may be wrong with WordPress.org or this server&#8217;s configuration.', 'performance-lab' );
+			wp_send_json_error( $status );
+		}
+
+		if ( ! $plugin_slug ) {
+			$status['errorMessage'] = esc_html__( 'Invalid plugin.', 'performance-lab' );
+			wp_send_json_error( $status );
+		}
+
+		// Install the plugin if it is not installed yet.
+		if ( ! isset( $plugins[ $plugin_basename ] ) ) {
+			// Replace new Plugin_Installer_Skin with new Quiet_Upgrader_Skin when output needs to be suppressed.
+			$skin     = new WP_Ajax_Upgrader_Skin( array( 'api' => $api ) );
+			$upgrader = new Plugin_Upgrader( $skin );
+			$result   = $upgrader->install( $api['download_link'] );
+
+			if ( is_wp_error( $result ) ) {
+				$status['errorMessage'] = $result->get_error_message();
+				wp_send_json_error( $status );
+			} elseif ( is_wp_error( $skin->result ) ) {
+				$status['errorMessage'] = $skin->result->get_error_message();
+				wp_send_json_error( $status );
+			} elseif ( $skin->get_errors()->has_errors() ) {
+				$status['errorMessage'] = $skin->get_error_messages();
+				wp_send_json_error( $status );
+			}
+		}
+
+		$result = activate_plugin( WP_PLUGIN_DIR . '/' . $plugin_basename );
```

Comment:
What if the plugin isn't located in `WP_PLUGIN_DIR` (even though this may be unlikely)? Also, the phpdoc for `activate_plugin()` says that this arg is _relative_ not absolute.

Reply:
Fantastic catch! I completely overlooked that aspect. Thank you for pointing it out. I've made the necessary updates, and you can find them in the commit here: https://github.com/WordPress/performance/pull/899/commits/2bd591d1eb3efd7172e1c356cc250df37616ec5a.